js/stdlib/ts/ALGO.ts

```diff
@@ -589,20 +589,83 @@ type QueryResult =
   | { succ: true, txn: any }
   | { succ: false, round: number }
 
-const doQuery = async (dhead:string, query: ApiCall<any>, pred: ((x:any) => boolean) = ((x) => { void(x); return true; })): Promise<QueryResult> => {
-  const res = await doQuery_(dhead, query);
-  const txns = res.transactions;
-  const ptxns = txns.filter(pred);
-  debug(dhead, {ptxns});
-
-  if ( ptxns.length == 0 ) {
-    return { succ: false, round: res['current-round'] };
+// ****************************************************************************
+// Event Cache
+// ****************************************************************************
+
+export const chooseMinRoundTxn = (ptxns: any[]) =>
+  ptxns.reduce((accum: any, x: any) =>
+      (x['confirmed-round'] < accum['confirmed-round']) ? x : accum, ptxns[0])
+
+export type RoundInfo = {
+  minRound?: number,
+  maxRound?: number,
+  specRound?: number,
+}
+
+export class EventCache {
+
+  cache: any[] = [];
+
+  currentRound = 0;
+
+  constructor() {
+    this.cache = [];
   }
 
-  const txn = ptxns.reduce((accum: any, x: any) =>
-    (x['confirmed-round'] < accum['confirmed-round']) ? x : accum, ptxns[0]);
-  return { succ: true, txn };
-};
+  async query(dhead: string, ApplicationID: number, roundInfo: RoundInfo, pred: ((x:any) => boolean)): Promise<QueryResult> {
+    const { minRound, maxRound, specRound } = roundInfo;
+    debug(dhead, `EventCache.query`, ApplicationID, minRound, specRound, maxRound, this.currentRound);
+
+    // Clear cache of stale transactions
+    const filterRound = minRound || specRound || 0;
+    this.cache = this.cache.filter(x => x['confirmed-round'] > filterRound);
+
+    // Check to see if the transaction we want is in cache
+    const initPtxns = this.cache.filter(pred);
```

Comment:
This is not quite right either, because imagine that we are looking for a call to m3 and m3 is called in a loop. The actual trace could be `m2 -> m3` (because `m2` is a timeout that goes back to the top of the while and enables another `m3`. In that case, the query we REALLY want is "Is m3 the earliest transaction?" Now, the only way this could happen is if the first `m3` (that we're looking for) is only valid in `[100, 200]` but it fails so `m2` happens at `201`, then the next `m3` happens at `202`. Your code is ignoring the maxRound specified, so presumably the real query is `m3 \in [100, 200]`, but you are doing `m3 \in [100, inf)` so you find the `202` one. You need to first filter the query up to the new minimum (not the specified), then compute the part of the txns that are valid for THIS query (the spec and max) and then see if the predicate applies